src/python/pants/backend/python/providers/python_build_standalone/rules.py

```diff
@@ -0,0 +1,244 @@
+class PBSPythonProvider(PythonProvider):
+    pass
+
+
+def _choose_python(
+    interpreter_constraints: InterpreterConstraints,
+    universe: Iterable[str],
+    pbs_versions: collections.abc.Collection[str],
+) -> str:
+    """Choose the highest supported patch of the lowest supported Major/Minor version."""
+    supported_python_triplets = interpreter_constraints.enumerate_python_versions(universe)
+    version_triplet: tuple[int, int, int] | None = None
+    for triplet in supported_python_triplets:
+        pbs_supported_version = ".".join(map(str, triplet)) in pbs_versions
+        if pbs_supported_version:
+            if version_triplet and version_triplet[:2] < triplet[:2]:
+                # This version is a major/minor above the previous supported one, we're done.
+                break
+
+            version_triplet = triplet
+
+    if version_triplet is None:
+        raise Exception(
+            softwrap(
+                f"""\
+                Failed to find a supported Python Build Standalone for Interpreter Constraint: {interpreter_constraints.description}
+
+                Supported versions are currently: {sorted(pbs_versions)}.
+
+                You can teach Pants about newer Python versions supported by Python Build Standalone
+                by setting the `known_python_versions` option in the {PBSPythonProviderSubsystem.name}
+                subsystem. Run `{bin_name()} help-advanced {PBSPythonProviderSubsystem.options_scope}`
+                for more info.
+                """
+            )
+        )
+
+    return ".".join(map(str, version_triplet))
+
+
+@rule
+async def get_python(
+    request: PBSPythonProvider,
+    python_setup: PythonSetup,
+    pbs_subsystem: PBSPythonProviderSubsystem,
+    platform: Platform,
+    named_caches_dir: NamedCachesDirOption,
+    cp: CpBinary,
+) -> PythonExecutable:
+    versions_info = pbs_subsystem.get_all_pbs_pythons()
+
+    python_version = _choose_python(
+        request.interpreter_constraints,
+        python_setup.interpreter_versions_universe,
+        versions_info,
+    )
+    pbs_py_info = versions_info[python_version][platform.value]
+
+    downloaded_python = await Get(
+        DownloadedExternalTool,
+        ExternalToolRequest(
+            DownloadFile(
+                pbs_py_info["url"],
+                FileDigest(
+                    pbs_py_info["sha256"],
+                    pbs_py_info["size"],
+                ),
+            ),
+            "python/bin/python3",
+        ),
+    )
+
+    await Get(
+        ProcessResult,
+        Process(
+            [
+                cp.path,
+                "--recursive",
+                "--no-clobber",
+                "python",
+                f"{PBS_SANDBOX_NAME}/{python_version}",
+            ],
+            level=LogLevel.DEBUG,
+            input_digest=downloaded_python.digest,
+            description=f"Install Python {python_version}",
+            append_only_caches=PBS_APPEND_ONLY_CACHES,
```

Comment:
Hmm, I feel the use of append-only/named-caches here isn't great, as I think it makes it incompatible with remote execution. In particular, even within a single session, we might run this on machine A to seed its named cache, and then execute downstream processes on machine B that don't have a seeded cache.

An alternative approach would be capturing digests and including them in sandboxes etc... but maybe this requires rearchitecting how the providers work.

So, definitely happy for it to be follow-up/filing a separate issue.

Reply:
(I lean heavily on remote caching by way of REAPI at $dayjob, but have not used remote *execution*. ) 

Some cross-linking notes:

 * This pattern is I think pre-existing at https://github.com/pantsbuild/pants/blob/69255bd97fa68f3a0f91b06330de59debbb7935b/src/python/pants/core/util_rules/adhoc_binaries.py#L119 and  https://github.com/pantsbuild/pants/blob/main/src/python/pants/backend/python/providers/pyenv/rules.py#L301https://github.com/pantsbuild/pants/blob/69255bd97fa68f3a0f91b06330de59debbb7935b/src/python/pants/backend/python/providers/pyenv/rules.py#L301
  * From simple search, #20485 appears to be providing a (golang) toolchain without `append_only_caches`
  * #20552 foundered on the non-optional use of named_caches
  * From the motivation for  #20552  I'm worried that naively treating the O(thousands) of little files in a Python install as digests.
  * Are named_caches used by REAPI at all today? #10078

Reply:
For some context, there is an REAPI-adajcent protocol called "Remote Asset API" which is designed to tell a remote execution server to fetch and download artifacts into a remote store. Pants does not implement it though.

Also, there is not append-only / named caches in remote execution unless a server operator takes actions to provide a persisted writeable directory to store that cache. If a sever operator does provide such a location, the Pants option `--remote-execution-append-only-caches-base-path` can be set to let Pants make use of it. Most of the servers implementing REAPI are not going to provide that out of the box.